src/Grpc.StatusProto/RpcStatusExtensions.cs

```diff
@@ -0,0 +1,210 @@
+// Copyright 2023 gRPC authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using Google.Protobuf;
+using Google.Protobuf.Reflection;
+using Google.Protobuf.WellKnownTypes;
+using Grpc.Core;
+
+namespace Grpc.StatusProto;
+
+/// <summary>
+/// Extensions for <see cref="Google.Rpc.Status"/> to retrieve detailed error information.
+/// Based on ideas from:
+/// https://github.com/googleapis/gax-dotnet/blob/main/Google.Api.Gax.Grpc/RpcExceptionExtensions.cs
+/// </summary>
+public static class RpcStatusExtensions
+{
+    /// <summary>
+    /// Retrieves the error details of type <typeparamref name="T"/> from the <see cref="Google.Rpc.Status"/>
+    /// message.
+    /// Note: experimental API that can change or be removed without any prior notice.
+    /// </summary>
+    /// <remarks>
+    /// <example>
+    /// For example, to retrieve any <see cref="Google.Rpc.ErrorInfo"/> that might be in the status details:
+    /// <code>
+    ///   var errorInfo = status.GetStatusDetail&lt;ErrorInfo&gt;();
+    ///   if (errorInfo is not null) {
+    ///      // ...
+    ///   }
+    /// </code>
+    /// </example>
+    /// </remarks>
+    /// <typeparam name="T">The message type to decode from within the error details.</typeparam>
+    /// <param name="status">The RPC status to retrieve details from. Must not be null.</param>
+    /// <returns>The first error details of type <typeparamref name="T"/> found, or null if not present</returns>
+    public static T? GetStatusDetail<T>(this Google.Rpc.Status status) where T : class, IMessage<T>, new()
+    {
+        var expectedName = new T().Descriptor.FullName;
```

Comment:
This is a bit of a hack. I've gotten a type's descriptors without creating an instance before. I'll see if I can find it.

Reply:
It was over service descriptors and I used reflection. I think I prefer this.

What about adding caching with a static `ConcurrentDictionary<Type, string>`? That reduces the allocation to the first time this method is used with each type.